lib/core/utils/selector-cache.js

```diff
@@ -0,0 +1,196 @@
+    if (
+      expressions[i].length > 1 &&
+      expressions[i].some(expression => isGlobalSelector(expression))
+    ) {
+      return;
+    }
+  }
+
+  const nodeSet = new Set();
+
+  // find nodes that match just a part of the selector in order
+  // to speed up traversing the entire tree
+  expressions.forEach(expression => {
+    // use the last part of the expression to find nodes as it's more
+    // specific. e.g. for `body h1` use `h1` and not `body`
+    const exp = expression[expression.length - 1];
+    let nodes = [];
+
+    // a complex selector is one that will require using
+    // matchesExpression to determine if it matches. these include
+    // pseudo selectors (:not), combinators (A > B), and any
+    // attribute value ([class=foo]).
+    let isComplexSelector =
+      expression.length > 1 || !!exp.pseudos || !!exp.classes;
+
+    if (isGlobalSelector(exp)) {
+      nodes = selectorMap['*'];
+    } else {
+      if (exp.id) {
+        // a selector must match all parts, otherwise we can just exit
+        // early
+        if (!selectorMap[idsKey] || !selectorMap[idsKey][exp.id]?.length) {
+          return;
+        }
+
+        // when using id selector (#one) we only find nodes that
+        // match the shadowId of the root
+        nodes = selectorMap[idsKey][exp.id].filter(
+          node => node.shadowId === shadowId
+        );
+      }
+
+      if (exp.tag && exp.tag !== '*') {
+        if (!selectorMap[exp.tag]?.length) {
+          return;
+        }
+
+        nodes = innerJoin(selectorMap[exp.tag], nodes);
+      }
+
+      if (exp.classes) {
+        if (!selectorMap['[class]']?.length) {
+          return;
+        }
+
+        nodes = innerJoin(selectorMap['[class]'], nodes);
+      }
+
+      if (exp.attributes) {
+        for (let i = 0; i < exp.attributes.length; i++) {
+          const attr = exp.attributes[i];
+
+          // an attribute selector that looks for a specific value is
+          // a complex selector
+          if (attr.type === 'attrValue') {
+            isComplexSelector = true;
+          }
+
+          if (!selectorMap[`[${attr.key}]`]?.length) {
+            return;
+          }
+
+          nodes = innerJoin(selectorMap[`[${attr.key}]`], nodes);
+        }
+      }
+    }
+
+    nodes.forEach(node => {
+      // for complex selectors we need to verify that the node
+      // actually matches the entire selector since we only have
+      // nodes that partially match the last part of the selector
+      if (isComplexSelector && !matchesExpression(node, expression)) {
+        return;
+      }
+
+      nodeSet.add(node);
+    });
+  });
+
+  // ie11 does not work with Array.from without a polyfill (missing
+  // `.entries`) but does have forEach. using a set saves time when
+  // wanting to make sure the same node isn't added twice (~3 seconds
+  // total over using array.includes)
+  let matchedNodes = [];
+  nodeSet.forEach(node => matchedNodes.push(node));
```

Comment:
If you're going to return an array, why bother with a set? Wouldn't it be easier to just use an array and then filter it for unique values before you return?